[SHK-30] Coupon 수량 감소 기능 구현

goseungwon opened the pull request:
### ⛏ 작업 사항
- 쿠폰을 지급시 쿠폰 수량이 감소한다.
- Custom Exception 으로 변경


### 📝 작업 요약
- 쿠폰 수량 기능 구현
- Exception 변경
- 쿠폰 발급 메서드 refactor

goseungwon committed:
Feat : [SHK-41] 쿠폰 발급 시점에 할당받은 사용자인지 검증한다.
- CouponService 에서 발급받은 member 인지 검증
- 이미 쿠폰을 발급 받은 사용자인 경우 예외처리

goseungwon committed:
Test : [SHK-40] 쿠폰 중복 발급 검증 테스트
- CouponServiceTest test 메서드 추가
- CouponEventControllerTest test 메서드 추가

goseungwon committed:
Test : [SHK-40] 쿠폰 중복 발급 검증 테스트
- CouponServiceTest test 메서드 추가

goseungwon committed:
Refactor : 쿠폰 중복 발급 exception 변경

변경 내용
1. DuplicateKeyException -> DuplicateRequestException

변경 이유
1. 적절한 에러로 변경

goseungwon committed:
Feat : 쿠폰 수량 감소 기능 구현

goseungwon committed:
Refactor : CouponEvent 필드 변경

변경 내용
1. Member -> MemberId, Coupon -> CouponId
2. 변경됨에 따른 코드 변경

변경 이유
1. 엔티티의 생명주기가 겹치지 않아서

goseungwon committed:
Refactor : CouponEvent 중복 확인 메서드 변경

변경 내용
1. findByMember -> existsByMemberIdAndCouponId

변경 이유
1. 효율성 증진

goseungwon committed:
Refactor : Custom Exception 생성

변경 내용
1. IllegalArgumentException -> OutOfStockException

변경 이유
1. 팀 컨벤션 변경

goseungwon committed:
Merge remote-tracking branch 'origin/develop' into feature/SHK-29

# Conflicts:
#	src/main/java/com/prgrms/kream/common/mapper/CouponEventMapper.java
#	src/main/java/com/prgrms/kream/domain/coupon/facade/CouponFacade.java
#	src/main/java/com/prgrms/kream/domain/coupon/repository/CouponEventRepository.java
#	src/main/java/com/prgrms/kream/domain/coupon/service/CouponEventService.java
#	src/test/java/com/prgrms/kream/domain/coupon/controller/CouponEventControllerTest.java
#	src/test/java/com/prgrms/kream/domain/coupon/service/CouponEventServiceTest.java

boompatron commented on src/main/java/com/prgrms/kream/common/exception/OutOfStockException.java:
```diff
@@ -0,0 +1,15 @@
+package com.prgrms.kream.common.exception;
+
+public class OutOfStockException extends RuntimeException {
```
제 의견이 반영됐군요!ㅎㅎ

goseungwon replied:
맞습니다 👏👏

boompatron commented on src/main/java/com/prgrms/kream/domain/coupon/dto/response/CouponEventResponse.java:
```diff
@@ -1,12 +1,9 @@
-		Member member
+		Long couponId,
+		Long memberId
 ) {
```
DTO에서 엔티티를 들고 있다가 ID를 가지고 있는 것으로 변경하셨군요. dto를 더 가볍게 하기 위함일까요?

goseungwon replied:
저번 스크럼 때 엔티티의 생명주기에 대해 이야기를 나눴는데, 엔티티의 생명주기가 다른 경우 id만 참조해서 끊어주는 것이 좋다고 판단했습니다.
[여기](https://www.youtube.com/watch?v=dJ5C4qRqAgA)도 참조했습니다!

boompatron commented on src/test/java/com/prgrms/kream/domain/coupon/controller/CouponEventControllerTest.java:
```diff
@@ -28,6 +29,7 @@ class CouponEventControllerTest extends MysqlTestContainer {
 	Jackson2ObjectMapperBuilder objectMapperBuilder;
 
 	@Test
+	@Transactional
```
테스트에도 transactional 어노테이션을 붙이셨군요! 혹시 이렇게 하면 얻을 수 있는 효과가 있나요??

goseungwon replied:
test 후 다른 test에 영향이 가지 않도록 롤백하기 위함입니다. [여기](https://www.inflearn.com/questions/257700/%EC%99%9C-test%EC%97%90-transactional%EC%9D%84-%EB%98%90-%EB%B6%99%EC%97%AC%EC%95%BC%ED%95%98%EB%82%98%EC%9A%94)를 확인해보시면 도움 될거라 믿습니다!


test 메서드에 `@Transactional`을 사용할 때는 주의할점이 하나 있습니다!
[이러한](https://javabom.tistory.com/103) 문제가 야기되기 대문에 service 메서드에 모두 `@Transactional`을 붙여줬습니다.

boompatron reviewed (approved):
구현하느라 수고하셨습니다! 몇 가지 질문 남겼습니다!

BnDC commented on src/main/java/com/prgrms/kream/domain/coupon/service/CouponEventService.java:
```diff
@@ -18,16 +18,17 @@
 			return toCouponEventResponse(savedCouponEvent);
 		}
 		throw new DuplicateRequestException("쿠폰을 중복으로 받을 수 없습니다.");
 	}
```
[early return]이 가능하다면 early return으로 작성하는 것이 좋을 거 같습니다.

[early return]: https://jheloper.github.io/2019/06/write-early-return-code/

goseungwon replied:
네 변경 했습니다!

BnDC commented on src/test/java/com/prgrms/kream/domain/coupon/service/CouponEventServiceTest.java:
```diff
@@ -13,15 +13,15 @@
 import com.prgrms.kream.domain.member.model.Member;
+import com.sun.jdi.request.DuplicateRequestException;
 
 @AutoConfigureMockMvc
```
ServiceTest에 `@AutoConfigureMockMvc` 를 사용한 이유가 있을까요?

goseungwon replied:
service, repository를 메모리에 올리려 했습니다!

BnDC reviewed (commented):
승원님 구현하느라 고생하셨습니다! 몇 가지 코멘트 남겼습니다

goseungwon committed:
Refactor : Early Return 사용

변경 내용
1. registerCouponEvent 메서드 구조 변경

변경 이유
1. 가독성 증가

goseungwon committed:
Test : Test 수정 및 추가

변경 내용
1. ServiceTest 에 불필요한 주석 제거
2. CouponTest 추가

변경 이유
1. 사용하지 않는 주석 제거
2. 쿠폰 수량에 대한 Test
